src/lib.rs

```diff
@@ -509,6 +511,107 @@ impl<'a, T: CBufItem, const SIZE: usize> CBufReader<'a, T, SIZE> {
+    ///
+    /// (i.e., [`fetch_next_item`](Self::fetch_next_item)
+    /// and [`available_items_iter`](Self::available_items_iter)).
+    #[inline]
+    pub fn set_next_index(&mut self, next: CBufIndex<SIZE>) {
+        self.next_local = next;
+    }
+
+    /// Finds the earliest valid index that makes the predicate true
+    /// assuming buffer contents are sorted by `predicate()` in `false`, `true` order.
+    ///
+    /// If `fromstart` is set, searches all available data, otherwise
+    /// the search starts at the `current_next_index()`.
+    ///
+    /// On completion, the `current_next_index()` is set to
+    /// the returned index, or to the end index if
+    /// no true value is found and InvalidIndexError.
+    /// if buffer contents are sorted by `predicate()` in `false`, `true` order.
+    ///
+    /// Returns `InvalidIndexError` if no such index exists.
+    ///
+    /// If data is not sorted by `predicate()` (i.e. if at least one
+    /// item that makes the predicate `false` is later than at
+    /// least one item that makes the predicate `true`) the
+    /// function may return either an index that makes the predicate
+    /// true, or an `InvalidIndexError`.
+    pub fn search(
+        &mut self,
+        predicate: &dyn Fn(T) -> bool,
+        fromstart: bool,
+    ) -> Result<CBufIndex<SIZE>, InvalidIndexError> {
+        let valids = self.current_valid_index_range();
+
+        // let failure = || {
+        //     self.set_next_index(valids.end);
+        //     Err(InvalidIndexError(()))
+        // };
+        // let success = | idx | {
+        //     self.set_next_index(idx);
+        //     Ok(idx)
+        // };
+        macro_rules! failure {
+            () => {{
+                self.set_next_index(valids.end);
+                Err(InvalidIndexError(()))
+            }};
+        }
+        macro_rules! success {
+            ($idx:ident) => {{
+                self.set_next_index($idx);
+                Ok($idx)
+            }};
+        }
+        if fromstart {
+            self.set_next_index(valids.start)
+        }
+        match self.fetch_next_item(false) {
+            ReadResult::None | ReadResult::SpinFail => return failure!(),
+            ReadResult::Skipped(v) | ReadResult::Success(v) => {
+                if predicate(v) {
+                    let idx = self.current_next_index() - 1;
+                    return success!(idx);
+                }
+            }
+        }
+        let mut ifalse = self.current_next_index() - 1;
+        let mut itrue = valids.end - 1;
+        // Verify that the predicate is true at the end
+        match self.fetch(itrue) {
+            Ok(v) => {
+                if !predicate(v) {
+                    return failure!();
+                }
+            }
+            Err(_) => return failure!(),
+        }
+        // At this point, ifalse < itrue and index to false and true values
+
+        while ifalse + 1 < itrue {
```

Comment:
Without knowing why you don't want `CBufIndex` to be orderable, I don't know know why `bisect` would be better.  

It looks like it might be hard to handle the error condition where the write pointer overtakes the binary search pointer, but calling that case an `InvalidIndexError` seems reasonable.

If the `CBufReader`'s `next_local` is sufficiently stale and `fromstart` is `false` the search will fail immediately if `range` starts with too much overwritten data.  I would begin the range at `next_local.max(self.current_valid_index_range().start)`, but once again, that requires `CBufIndex` to be orderable.


(Looking at the `Range` documentation, it doesn't appear to have the ability to do the intersection of two `Range`s which would a way of disguising the fact that it is using the ordering of the indices.)

Reply:
Between discussions with you and thinking through it, I'm OK with the comparison operators here.

Reply:
Apart from removing the commented-out code at [line 576], `search` is good.

[line 576]: https://github.com/BlackCAT-CubeSat/bc-buf/pull/3/files#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R576